gnovm/pkg/gnolang/machine.go

```diff
@@ -1820,20 +1855,40 @@ func (m *Machine) LastCallFrame(n int) *Frame {
 			}
 		}
 	}
-	panic("frame not found")
+
+	if !safe {
+		panic("frame not found")
+	}
+
+	return nil
 }
 
 // pops the last non-call (loop) frames
 // and returns the last call frame (which is left on stack).
-func (m *Machine) PopUntilLastCallFrame() *Frame {
+func (m *Machine) PopUntilLastCallFrame(safe bool) *Frame {
 	for i := len(m.Frames) - 1; i >= 0; i-- {
-		fr := &m.Frames[i]
+		fr := m.Frames[i]
 		if fr.Func != nil || fr.GoFunc != nil {
 			// TODO: optimize with fr.IsCall
 			m.Frames = m.Frames[:i+1]
 			return fr
 		}
+
+		fr.Popped = true
+	}
+
+	// The frame wasn't found. If not in safe mode, panic to avoid any potential
+	// nil pointer dereferences.
+	if !safe {
+		panic("no last call frame found")
 	}
+
+	// In safe mode we return nil and no frames are popped, so update the frames' popped flag.
+	// This is expected to happen infrequently.
+	for _, frame := range m.Frames {
+		frame.Popped = false
+	}
```

Comment:
Can you explain why this is important?  Because it completely changes the behavior of the function.

Reply:
I will have to look and get back to you on this. It also looks odd to me which means it's wrong or I need to document it better

Reply:
I modified this to remove the old and unnecessary functionality.  Now the only real change is that frames have their `Popped` flag marked when they are popped from the stack.